h2o-py/h2o/backend/connection.py

```diff
@@ -471,49 +471,52 @@ def request(self, endpoint, data=None, json=None, filename=None, save_to=None):
-            resp = requests.request(method=method, url=url, data=request_data, json=json, params=params,
-                                    stream=stream, **args)
-            if isinstance(save_to, types.FunctionType):
-                save_to = save_to(resp)
-            self._log_end_transaction(start_time, resp)
-            return self._process_response(resp, save_to)
-
-        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as e:
-            if self._local_server and not self._local_server.is_running():
-                self._log_end_exception("Local server has died.")
-                raise H2OConnectionError("Local server has died unexpectedly. RIP.")
-            else:
+            # Make the request
+            start_time = time.time()
+            try:
+                self._log_start_transaction(endpoint, request_data, json, filename, params)
+                args = self._request_args()
+                resp = requests.request(method=method, url=url, data=request_data, json=json, params=params,
+                                        stream=stream, **args)
+                if isinstance(save_to, types.FunctionType):
+                    save_to = save_to(resp)
+                self._log_end_transaction(start_time, resp)
+                return self._process_response(resp, save_to)
+
+            except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as e:
+                if self._local_server and not self._local_server.is_running():
+                    self._log_end_exception("Local server has died.")
+                    raise H2OConnectionError("Local server has died unexpectedly. RIP.")
+                else:
+                    self._log_end_exception(e)
+                    raise H2OConnectionError("Unexpected HTTP error: %s" % e)
+            except requests.exceptions.Timeout as e:
                 self._log_end_exception(e)
-                raise H2OConnectionError("Unexpected HTTP error: %s" % e)
-        except requests.exceptions.Timeout as e:
-            self._log_end_exception(e)
-            elapsed_time = time.time() - start_time
-            raise H2OConnectionError("Timeout after %.3fs" % elapsed_time)
-        except H2OResponseError as e:
-            err = e.args[0]
-            if isinstance(err, H2OErrorV3):
-                err.endpoint = endpoint
-                err.payload = (request_data, json, filename, params)
-            raise
+                elapsed_time = time.time() - start_time
+                raise H2OConnectionError("Timeout after %.3fs" % elapsed_time)
+            except H2OResponseError as e:
+                err = e.args[0]
+                if isinstance(err, H2OErrorV3):
+                    err.endpoint = endpoint
+                    err.payload = (request_data, json, filename, params)
+                raise
+        finally:
+            if hasattr(request_data, "close"):
+                request_data.close()
```

Comment:
The only change is the try-finally where I make sure to close an open file. I will extract this to a separate PR. Now I keep it here to be able to use [PYTHONDEVMODE=1](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONDEVMODE) without having too much "false" alarms.